tests/test_recovery_token_flow.py

```diff
@@ -59,3 +63,89 @@ def test_recovery_token_authenticated_user(
         },
     )
     assertRedirects(response, reverse("admin:index"))
+
+
+@pytest.mark.django_db
+def test_recovery_token_invalid_code_shows_error(client, user, totp_device):
+    recovery_url = reverse("maykin_2fa:recovery")
+
+    client.post(
+        reverse("admin:login"),
+        data={
+            "admin_login_view-current_step": "auth",
+            "auth-username": "johny",
+            "auth-password": "password",
+        },
+    )
+
+    response = client.get(recovery_url)
+    assert response.status_code == 200
+    assertTemplateUsed(response, "maykin_2fa/recovery_token.html")
+
+    response = client.post(
+        recovery_url,
+        data={
+            "admin_login_view-current_step": "backup",
+            "backup-otp_token": "INVALID-CODE",
+        },
+    )
+
+    assert response.status_code == 200
+    assertTemplateUsed(response, "maykin_2fa/recovery_token.html")
+    assertContains(
+        response, "Invalid token. Please make sure you have entered it correctly."
+    )
+
+
+@pytest.fixture
+def backup_tokens_device(user):
+    return StaticDevice.objects.create(user=user, name="backup")
+
+
+@pytest.mark.django_db
+def test_custom_backup_token_form_invalid_token_raises_custom_error_code(
+    user, backup_tokens_device
+):
+    form = CustomBackupTokenForm(
+        user=user,
+        data={"otp_token": "invalid-code"},
+        initial_device=backup_tokens_device,
+    )
+
+    assert not form.is_valid()
+
+    errors = form.errors.get("otp_token")
+    assert errors
+
+    assert "Invalid token" in errors[0]
+
+
+@pytest.mark.django_db
+def test_clean_otp_token_raises_validation_error_with_custom_code(user):
+    form = CustomBackupTokenForm(
+        user=user,
+        data={"otp_token": "invalid-code"},
+        initial_device=backup_tokens_device,
+    )
+    form.full_clean()
+
+    errors = form.errors.get("otp_token")
+    assert errors
+    assert "Invalid token" in errors[0]
+
+
+@pytest.mark.django_db
+def test_clean_otp_token_raises_validation_error_with_code(user):
+    form = CustomBackupTokenForm(
+        user=user,
+        data={"otp_token": "invalid-code"},
+        initial_device=backup_tokens_device,
+    )
+
+    form.cleaned_data = {"otp_token": "invalid-code"}
+
+    with pytest.raises(ValidationError) as exc_info:
+        form.clean_otp_token()
+
+    assert exc_info.value.message == "Invalid token"
+    assert exc_info.value.code == "invalid_token"
```

Comment:
I appreciate the tests! But I think the test above `test_recovery_token_invalid_code_shows_error` is enough to verify that the issue has been fixed, it's also the closest to the way the user would interact with the form, so I think we can remove the other (more specific) tests

```suggestion
```